crypto/evp/evp_fetch.c

```diff
@@ -51,6 +53,67 @@ struct method_data_st {
     void (*destruct_method)(void *method);
 };
 
+static int add_names_to_namemap(OSSL_NAMEMAP *namemap,
+                                const char *names)
+{
+    const char *p, *q;
+    size_t l;
+    int id = 0;
+
+    /* Check that we have a namemap and that there is at least one name */
+    if (namemap == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    /*
+     * Check that no name is an empty string, and that all names have at
+     * most one numeric identity together.
+     */
+    for (p = names; *p != '\0'; p = (q == NULL ? p + l : q + 1)) {
+        int this_id;
+
+        if ((q = strchr(p, NAME_SEPARATOR)) == NULL)
+            l = strlen(p);       /* offset to \0 */
+        else
+            l = q - p;           /* offset to the next separator */
+
+        this_id = ossl_namemap_name2num_n(namemap, p, l);
+
+        if (*p == '\0' || *p == NAME_SEPARATOR) {
+            ERR_raise(ERR_LIB_EVP, ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+        if (id == 0)
+            id = this_id;
+        else if (this_id != 0 && this_id != id) {
+            ERR_raise(ERR_LIB_EVP, ERR_R_INTERNAL_ERROR);
```

Comment:
Again - could this be an error from a provider, i.e. not an internal error? Or in fact inconsistent data from two different providers? This actually implies that providers must not have algorithm synonyms as two separate entries in an OSSL_ALGORITHM array, e.g. this is an error:

````
    { "dhKeyAgreement", "default=yes", dh_keyexch_functions },
    { "dh", "default=yes", dh_keyexch_functions },
````
Because, if another provider defines it like this it will fail:
````
    { "dh:dhKeyAgreement", "default=yes", dh_keyexch_functions },
````

Feasibly, we could actually add a check from such synonymous entries?

Reply:
That...  could be done, sure.  So question is, do we want to allow that form of synonym or disallow it?

Reply:
I tend towards disallow, to encourage a consistent way of doing things across all providers.

Reply:
Hmmm, that's going to take an effort that doesn't quite belong here.  Do you mind if it happens in a separate PR?

Reply:
Separate PR is fine.